File: benchmark/python/merge.py
import time
arr = list(range(1000, 0, -1))
start = time.time() * 1000
n = len(arr)
width = 1
while width < n:
    for left in range(0, n, 2 * width):
        mid = left + width - 1
        if mid < n - 1:
            right = left + 2 * width - 1
            if right >= n:
                right = n - 1
            n1 = mid - left + 1
            n2 = right - mid
            L = []
            R = []
            for k in range(n1):
                L.append(arr[left + k])
            for k in range(n2):
                R.append(arr[mid + 1 + k])
            i = j = 0
            m = left
            while i < n1 and j < n2:
                if L[i] <= R[j]:
                    arr[m] = L[i]
                    i += 1
                else:
                    arr[m] = R[j]
                    j += 1
                m += 1
            while i < n1:
                arr[m] = L[i]
                i += 1
                m += 1
            while j < n2:
                arr[m] = R[j]
                j += 1
                m += 1
    width *= 2
end = time.time() * 1000
print(f"First: {arr[0]}")
print(f"Last: {arr[-1]}")
print(f"Time: {end - start:.0f}ms")

File: benchmark/lua/merge.lua
local start = os.clock()
local arr = {}
for i = 1000, 1, -1 do table.insert(arr, i) end
local n = #arr
local width = 1
while width < n do
    local left = 1
    while left <= n do
        local mid = left + width - 1
        if mid < n then
            local right = left + 2 * width - 1
            if right > n then right = n end
            local n1 = mid - left + 1
            local n2 = right - mid
            local L = {}
            local R = {}
            for k = 1, n1 do L[k] = arr[left + k - 1] end
            for k = 1, n2 do R[k] = arr[mid + k] end
            local i, j = 1, 1
            local m = left
            while i <= n1 and j <= n2 do
                if L[i] <= R[j] then
                    arr[m] = L[i]
                    i = i + 1
                else
                    arr[m] = R[j]
                    j = j + 1
                end
                m = m + 1
            end
            while i <= n1 do
                arr[m] = L[i]
                i = i + 1
                m = m + 1
            end
            while j <= n2 do
                arr[m] = R[j]
                j = j + 1
                m = m + 1
            end
        end
        left = left + 2 * width
    end
    width = width * 2
end
local finish = os.clock()
print("First: " .. arr[1])
print("Last: " .. arr[n])
print(string.format("Time: %.0fms", (finish - start) * 1000))

